### ISSUE TITLE
[Question]

### ISSUE THREAD

**Issue by admin-elvistrujanovic** [JingMatrix/ChromeXt#73]
Is there any project for rooted users to install Chrome extensions support to their chromium based browsers like Google Chrome, Microsoft Edge, etc... using LSPosed/Xposed Framework or if it doesn't exist, can it be added to this module so we can have native support of chrome extensions inside any chromium-based browser we can have on our devices.

**Comment by MohamadBayz** [comment]
Kiwi Browser has support for extension also i think it's source code can be found in GitHub

**Comment by admin-elvistrujanovic** [comment]
> Kiwi Browser has support for extension also i think it's source code can be found in GitHub

Yeah but they do not support injecting patches into for example official Google Chrome. Imagine dreams come true so that we can in official Google Chrome or Microsoft Edge gets full chrome extensions support. Not Kiwi Browser, not custom browsers with built-in support - just direct **native** support inside official browsers using LSPosed/LSPatch/Xposed Framework 🥰

**Comment by MohamadBayz** [comment]
Ik, i was just suggesting ideas to be beneficial for the dev to look into

**Comment by JingMatrix** [comment]
Thanks for your passionate question.
I once wished so as well. Here are some related questions I have investigated before, which might be helpful for you to understand the situation.

Q: Why extensions are not supported in mobile browsers?
A: Chromium project explicitly disabled it. The related code of extensions is avaiable in the Chroimum project, but is excluded from the build processs of Android Chrome. Somehow, Kiwi adds it back to the build process.

Q: Can we learn from Kiwi and add the extension code back to the build process using Xposed framwork?
A: Barely possible. Xposed framework can hook Java codes, but barely C/C++ codes. And it requires some works on the C/C++ codes to add the extension code back to the build process.

Q: What about rewriting a new implementation of Chrome extensions?
A: Currently, it is difficult. One need to re-implement the whole set of [Browser extension APIs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions).

Q: What is the specific difficulty compared to re-implementing the UserScripts APIs?
A: Running scripts in background. ChromeXt is quite strong with the help of DevTools, but still couldn't handle this problem. Extensions exist mainly for the reason of background scripts.

It is more likely that the Chromium project would change their minds to support extensions for mobile browsers before I could finish the re-implementation.
However, if I have ideas for running scripts in backgroud, the remained stuff won't be difficult anymore.

**Comment by luxysiv** [comment]
@JingMatrix Can you support this app : https://play.google.com/store/apps/details?id=org.triple.banana

**Comment by JingMatrix** [comment]
> @JingMatrix Can you support this app : https://play.google.com/store/apps/details?id=org.triple.banana

Please open a new issue.

**Comment by MohamadBayz** [comment]
Mhmm quite interesting, so we can't Patch google chrome either to implement the extension API? Because lspatch also rely on Xposed framework
But what about Revanced they don't rely on Xposed framework is it difficult to do the same like Revanced?

**Comment by JingMatrix** [comment]
In the setting of current issue, there is no difference between Revanced and Xposed framework. Simply because reverse engineering C/C++ is much harder than reverse engineering Java.

**Comment by admin-elvistrujanovic** [comment]
> Thanks for your passionate question.
> I once wished so as well. Here are some related questions I have investigated before, which might be helpful for you to understand the situation.
> 
> Q: Why extensions are not supported in mobile browsers?
> A: Chromium project explicitly disabled it. The related code of extensions is avaiable in the Chroimum project, but is excluded from the build processs of Android Chrome. Somehow, Kiwi adds it back to the build process.
> 
> Q: Can we learn from Kiwi and add the extension code back to the build process using Xposed framwork?
> A: Barely possible. Xposed framework can hook Java codes, but barely C/C++ codes. And it requires some works on the C/C++ codes to add the extension code back to the build process.
> 
> Q: What about rewriting a new implementation of Chrome extensions?
> A: Currently, it is difficult. One need to re-implement the whole set of [Browser extension APIs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions).
> 
> Q: What is the specific difficulty compared to re-implementing the UserScripts APIs?
> A: Running scripts in background. ChromeXt is quite strong with the help of DevTools, but still couldn't handle this problem. Extensions exist mainly for the reason of background scripts.
> 
> It is more likely that the Chromium project would change their minds to support extensions for mobile browsers before I could finish the re-implementation.
> However, if I have ideas for running scripts in backgroud, the remained stuff won't be difficult anymore.

Nice - I'm ready to wait as long as it requires for the module/tool for patching the official browsers because I currently believe after many many years, Google just don't want and not just Google, the whole Chromium don't want Chrome extensions to be able to mobile devices so I believe that they will never add it. Especially because of Adblockers and many other extensions that exist on desktop devices can be found on Android/iOS devices as an alternative app or something.

So there is many reasons why they will probably never add an official support but with your work of implementation (patching methods like mentioned above or something similar to that), this could be changed unofficially to official without forcing them to support mobile devices like desktop devices.


Even Firefox stopped native support of extensions that was used for desktop and make something similar to be only for mobile devices because of safety and etc... and that also says that they literally all working on limiting somehow mobile devices from desktop features. But the most reasons are security and safety for their customers/users.


So only unofficial projects can give to us an official support on mobile devices we all want and deserve ❤️🥰🫂


Happy work our dear dev! And good luck with the revolution of chrome extensions support ✨

**Comment by MohamadBayz** [comment]
I did a quick digging o found this
https://github.com/wchen342/chromium-android-extension
Are they useable? They're for Chromium i don't know if it's reliable for google chrome

**Comment by JingMatrix** [comment]
Thanks for your searches. Basically, you can imagine that the Kiwi project as an updating version of what you found.
The project adds extension codes back to the build process of Android Chrome.

**Comment by MohamadBayz** [comment]
Oh, ok thanks for checking

**Comment by JingMatrix** [comment]
Now I get an idea of supporting background scripts.
In the future, `ChromeXt` will allow users to install WebExtentions.

Be patient, the plan is to re-implement the whole set of WebExtentions APIs.
You can keep track of [development plans](https://github.com/JingMatrix/ChromeXt#development-plans) to see how much is / will be done of supporting  WebExtentions.

I close this issue since the answer is clearly given.

**Comment by admin-elvistrujanovic** [comment]
> Now I get an idea of supporting background scripts.
> In the future, `ChromeXt` will allow users to install WebExtentions.
> 
> Be patient, the plan is to re-implement the whole set of WebExtentions APIs.
> You can keep track of [development plans](https://github.com/JingMatrix/ChromeXt#development-plans) to see how much is / will be done of supporting  WebExtentions.
> 
> I close this issue since the answer is clearly given.

wait... WebExtensions aka DevTool Extensions will be only supported? There's no way to implement the whole support of Chrome Extensions API and support into the module? sorry if I in wrong way understand this.

**Comment by JingMatrix** [comment]
No, you misunderstood. WebExtension is the standard notion for what you call browser extensions.

**Comment by admin-elvistrujanovic** [comment]
> No, you misunderstood. WebExtension is the standard notion for what you call browser extensions.

Oh so... you mean that Chrome API and Chrome Extension will be supported (and that includes Chrome Web Store) 🥰 - Wow! Internet just make me to misunderstood due to reading about thing called WebExtension by mozilla.org but those docs requires modifications by dev to port extension to WebExtension to work but it's good to know that in this case, that is not the same 🫂 - Thank you dev so much for the future of Chrome Extensions on official chromium browsers 🥰🫂
